File: hw/bonilla/U3/HW23Strategy/Strategyjs/controller/QuickSort.js
const SortingStrategy = require('./SortingStrategy');

class QuickSort extends SortingStrategy {
    sort(data) {
        console.log("Usando Quick Sort...");
        const sortedData = [...data];
        this.quickSort(sortedData, 0, sortedData.length - 1);
        this.mostrarArreglo(sortedData);
        return sortedData;
    }

    quickSort(arr, low, high) {
        if (low < high) {
            const pi = this.particion(arr, low, high);
            this.quickSort(arr, low, pi - 1);
            this.quickSort(arr, pi + 1, high);
        }
    }

    particion(arr, low, high) {
        const pivot = arr[high];
        let i = low - 1;

        for (let j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
        }

        [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
        return i + 1;
    }

    mostrarArreglo(data) {
        process.stdout.write("Resultado: ");
        data.forEach(num => process.stdout.write(num + " "));
        console.log();
    }
}

module.exports = QuickSort;

File: hw/bonilla/U3/HW23Strategy/Strategy/src/main/java/ec/edu/espe/strategy/controller/QuickSort.java
package ec.edu.espe.strategy.controller;

/**
 *
 * @author Bonilla David SoftCrafters
 */
public class QuickSort implements SortingStrategy {

    @Override
    public int[] sort(int data[]) {
        System.out.println("Usando Quick Sort...");
        int[] sortedData = data.clone();
        quickSort(sortedData, 0, sortedData.length - 1);
        mostrarArreglo(sortedData);
        return sortedData;
    }

    private void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = particion(arr, low, high);

            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    private int particion(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;

                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    private void mostrarArreglo(int[] data) {
        System.out.print("Resultado: ");
        for (int num : data) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}
